java/src/test/java/org/wildfly/openssl/BasicOpenSSLCustomEngineTest.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.openssl;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ * @author Heyuan Liu
+ */
+public class BasicOpenSSLCustomEngineTest {
+
+    @Test
+    public void testUnknownEngine() {
+        System.setProperty(SSL.ORG_WILDFLY_OPENSSL_ENGINE, "unknown");
```

Comment:
Each test case should also reset this system property to its original value in a `finally` block.

Reply:
Okay. I will fix this.

Reply:
@fjuma When I set the engine to qat, I found a performance issue caused by ByteBuffer. You can get the detail info from this project [ByteBufferPerformance](https://github.com/heyuanliu-intel/ByteBufferPerformance).

And this is the [Thread Dump File](https://raw.githubusercontent.com/heyuanliu-intel/AirliftPerformance/main/threaddump/thread1.txt).

So my fix is change from the direct byte buffer allocation to byte buffer pool.
`ByteBuffer buf = ByteBuffer.allocateDirect(len);`
to 
`  DefaultByteBufferPool.PooledByteBuffer pool = DefaultByteBufferPool.DIRECT_POOL.allocate();
   ByteBuffer buf = pool.getBuffer();
   buf.limit(len);`

So the CPU utilization can boost from 10% to 50%, and the performance can boost 5X.

Should I open a new PR or just add this code change commit to this PR?

Reply:
https://github.com/heyuanliu-intel/wildfly-openssl/blob/31d55d095b54b2469e21256926b8ba16d6ca85a9/java/src/main/java/org/wildfly/openssl/OpenSSLEngine.java#L243

https://github.com/heyuanliu-intel/wildfly-openssl/blob/31d55d095b54b2469e21256926b8ba16d6ca85a9/java/src/main/java/org/wildfly/openssl/OpenSSLEngine.java#L257-L276

Reply:
DefaultByteBufferPool.PooledByteBuffer pool = DefaultByteBufferPool.DIRECT_POOL.allocate();
            ByteBuffer buf = pool.getBuffer();
            buf.limit(len);
            try {
                final long addr = memoryAddress(buf);

                src.limit(pos + len);

                buf.put(src);
                src.limit(limit);

                sslWrote = SSL.getInstance().writeToSSL(ssl, addr, len);
                if (sslWrote > 0) {
                    src.position(pos + sslWrote);
                    return sslWrote;
                } else {
                    src.position(pos);
                }
            } finally {
                pool.close();
            }

Reply:
The same code change should be applied to all the ByteBuffer allocateDirect.

Reply:
@heyuanliu-intel Please create a separate [WFSSL issue](https://issues.redhat.com/projects/WFSSL/issues) and a separate PR for this and we can review that more closely and discuss there. Thanks!